### PR TITLE
Merge in acl-anthology-py

### PR DESCRIPTION
**Pull request by mbollmann**
This merges https://github.com/mbollmann/acl-anthology-py/ into this repo under a new folder `python/`.

The aim is for this to be a permanent move, i.e., the current acl-anthology-py will be archived after the successful merge, and all development should continue here. That's also why I chose to merge the entire history of that repo, because I believe it's useful not to lose it.

I assume that it will take some tweaking to get all workflows and integrations to work correctly. I was also quite opinionated about the pre-commit config and other automations while working on this myself, so we'll have to see what we'd like to keep and what should be adjusted.

### CODE REVIEW

**Comment by mbollmann** [comment]
NB: Build's failing because black & ruff are currently running in two different places, with different configuration settings.

File: python/.gitignore
Comment (akoehn) [inline python/.gitignore:1]: This should be toplevel imo

File: python/LICENSE.header
Comment (akoehn) [inline python/LICENSE.header:1]: What is this file for?

**Review by akoehn — COMMENTED** [review]
Onl had a few minutes. Is this all new implementation or is it derived from previous code? In that case the copyright noticed should be carried over.

File: python/LICENSE.header
Comment (mbollmann) [reply python/LICENSE.header:1]: This is what gets placed at the top of new files. Although with more than one contributor, maybe we want to do that manually now.

**Comment by mbollmann** [comment]
> Is this all new implementation or is it derived from previous code? In that case the copyright noticed should be carried over.

Both. I looked at copyright notices in the original files but it was mostly just my name up there, I think we weren't very diligent with keeping those updated. I can double-check though that I haven't missed anything.

**Comment by mbollmann** [comment]
Some points that might need to be resolved:

- I've put black and ruff (and some other things) in the pre-commit configuration; I don't really like that we're invoking them separately in the Makefile here. If we want to merge this, I guess we need to decide on one solution. Is your _(=&nbsp;plural)_ opinion still that you don't want hooks that modify files in the pre-commit config? I don't see the point of not having them simply in pre-commit; giving the large amount of vacuous "run black" commits that we're constantly seeing, I feel it would be beneficial to have these run automatically.
- My pyproject config targets Python 3.10–3.12; ours still targets 3.7–3.8. I wonder if we want to bump this up globally, potentially requiring changes in random `bin/` scripts from black/ruff,  or if we want to make a distinction here between the `bin/` scripts and the library.
- I've played around with using [Codecov](https://app.codecov.io/gh/mbollmann/acl-anthology-py) for test coverage and [Read the Docs](https://acl-anthology-py.readthedocs.io/en/latest/) for documentation. I believe both of these require granting them permissions on the repo, which I'm not allowed to do. Either way, we'd have to make a decision here.
- You see that there are a lot of new automated CI checks, targeting different OSes and Python versions. I believe that's a good thing, but I also wonder if Github workflows can be configured so that the Python checks only run (and are only required) on changes to the Python library, and maybe even that not every change in the Python library requires a full build of the website + preview. This last point is mainly what's kept me from preparing this merge earlier, as I fear that development of this library (and merging in new features etc.) might get more cumbersome if it's within this repo. One solution might be to use a separate `python-dev` branch that Python-library features get merged into, and only run check-build/preview upon a PR into `master`; but I haven't full thought this through yet.

**Comment by akoehn** [comment]
I'm fine with hooks changing the code.

I think we can bump the python requirement to 3.10.

It is trivial to make nly run coverage checks on changes in the subdir: https://stackoverflow.com/a/63822945

**Comment by akoehn** [comment]
Forgot the third: what is the advantage of codecoclv, given that

> How does pricing work? Only paid users can view reports and post statuses/comments.

**Comment by mbollmann** [comment]
> Forgot the third: what is the advantage of codecoclv, given that
> 
> > How does pricing work? Only paid users can view reports and post statuses/comments.

I'm not sure what that refers to? Codecov provides a graphical overview of test coverage (as can be [seen here for my acl-anthology-py repo](https://app.codecov.io/gh/mbollmann/acl-anthology-py)) as well as CI integration by posting reports about coverage changes in PRs and such ([example here](https://github.com/mbollmann/acl-anthology-py/pull/11#issuecomment-1773856673), entirely configurable of course). None of this requires a paid plan. (cf. https://about.codecov.io/pricing/)

There may be other providers offering something similar, I just happened to know Codecov.

**Comment by akoehn** [comment]
The quote was from the codecov GitHub plugin, but that probably only refers to private repos.

**Comment by mjpost** [comment]
Seconding @akoehn's comments above. Python 3.7 is [EOL'd](https://devguide.python.org/versions/) and anyway our user base is small enough that we can be progressive. Also happy to get plugged into readthedocs etc. We can update scripts lazily. Let's merge before it gets too out-of-date? Is there anything in the way?

**Comment by mbollmann** [comment]
@mjpost I can go over it one more time to see if I missed anything I wanted to address, but yeah, it would be great to not wait too long with merging. It's not that it will get out-of-date, but rather that I can't continue working on it before this is merged, since I changed the folder structure from my original repo. :)

I can also message you what app would need to get authorized where for the integrations to work.

**Review by mjpost — APPROVED** [review]

**Comment by mjpost** [comment]
I've approved this, but I wonder if I should port one script over first (e.g., an ingestion script) to get a feel for the new codebase. I'm not sure what your timeline is here. It looks like all the old code is still here redundantly, so as we move scripts over we could start deleting slowly, or just leave it there for backward compatibility.

One thought: instead of a `python` directory, isn't the convention usually to have the module name as the top-level directory? Also, I see unfortunately that the [acl-anthology](https://pypi.org/project/acl-anthology/) pypi package name is taken. I wonder if we could ask that guy to turn it over to us, since it seems to be abandoned.

**Comment by mjpost** [comment]
A great trial script to update would be `bin/write_bibkeys_to_xml.py`. This one currently has to be run manually and takes some time. I'd love to be able to use an API for it and incorporate it directly into the ingestion script. Just mentioning it in case you want to take a stab (since you wrote it originally and it's fairly small).

**Comment by mbollmann** [comment]
> I've approved this, but I wonder if I should port one script over first (e.g., an ingestion script) to get a feel for the new codebase. I'm not sure what your timeline is here. It looks like all the old code is still here redundantly, so as we move scripts over we could start deleting slowly, or just leave it there for backward compatibility.

I was assuming that after this merge, I would port the build pipeline first and foremost, which might reveal a few still missing pieces of functionality in the new library. I wish I had a timeline — I'm completely buried in work for the foreseeable future — but maybe the upcoming NLP-OSS workshop will give me some headspace to finish up this merge at least.

> One thought: instead of a `python` directory, isn't the convention usually to have the module name as the top-level directory?

The module resides in `python/acl_anthology/`, so the directory is named after the module name.

> Also, I see unfortunately that the [acl-anthology](https://pypi.org/project/acl-anthology/) pypi package name is taken. I wonder if we could ask that guy to turn it over to us, since it seems to be abandoned.

Yes, that's why it's `acl-anthology-py` on PyPI. I haven't attempted to contact the person yet, but we certainly could.

**Comment by mbollmann** [comment]
FWIW, an admin of the acl-org organization would need to authorize [Codecov](https://app.codecov.io/gh/) to access this repo, for that integration to work.

**Comment by mbollmann** [comment]
I believe this is ready for merging.

I took the liberty of updating our top-level README with information about the Python package, and also some status badges that I thought would be nice to have.

There is a randomly failing build check that is probably just a fluke; I'll see if I can get it to pass.

Steps after the merge:

- [ ] Give Codecov.io access to this repo @mjpost
- [ ] Change readthedocs configuration to build from this repo @mbollmann 
- [ ] Add a note to the old repo and archive it @mbollmann

**Comment by mbollmann** [comment]
Any objections to merging this before the holidays? Particularly with the README that I updated since you last reviewed?

**Review by akoehn — APPROVED** [review]
All good from my side

**Comment by mjpost** [comment]
Same, would be good to incorporate this.

**Comment by mjpost** [comment]
All right!

**Comment by akoehn** [comment]
🎉🥳 a gift at the right time :-)

**Pull request merged**
